Limit the size of served files

pietroalbini opened the pull request:
This PR limits how big the files we serve from docs.rs are, to avoid OOMs. The current limit is 5MB, but that can be changed by tweaking a constant in the source code.

This PR adds limits when downloding the files from S3 and the database, and when decompressing previously downloaded files. For now the user will see a "resource not found" error page, but a dedicated page can be created in the future.

This PR also refactors compression a bit to better enable testing. It's best to review the PR commit-by-commit.

r? @jyn514

Kixiron commented on src/utils/sized_buffer.rs:
```diff
@@ -0,0 +1,72 @@
+        }
+    }
+
+    pub(crate) fn reserve(&mut self, amount: usize) {
+        if self.inner.len() + amount > self.limit {
+            self.inner.reserve_exact(self.limit - self.inner.len());
+        } else {
+            self.inner.reserve_exact(amount);
+        }
+    }
```
Maybe this should return the amount that it ended up reserving so that we can pivot off of that as needed, allowing us to maybe less-fatally operate by retrying writes after the buffer's flushed

pietroalbini replied:
Do you see this being needed anywhere in the current PR? If future code needs it, the return value can be added in a future PR.

Kixiron commented on src/storage/mod.rs:
```diff
@@ -324,12 +324,12 @@ mod test {
 
-            let file = backend.get("rustdoc/Cargo.toml").unwrap();
+            let file = backend.get("rustdoc/Cargo.toml", std::usize::MAX).unwrap();
             assert_eq!(file.content, b"data");
             assert_eq!(file.mime, "text/toml");
             assert_eq!(file.path, "rustdoc/Cargo.toml");
 
-            let file = backend.get("rustdoc/src/main.rs").unwrap();
+            let file = backend.get("rustdoc/src/main.rs", std::usize::MAX).unwrap();
```
I think these tests could be improved by setting the limits to under, over and exactly at the length of the content we give, that way we can ensure that the buffering works with actual files as well

pietroalbini replied:
Hmm, these tests are not testing the size limits, they're just the older tests I had to change due to the new argument.

Kixiron commented on src/web/file.rs:
```diff
@@ -5,12 +5,14 @@ use crate::{db, error::Result};
 use postgres::Connection;
 
+const MAX_FILE_SIZE: usize = 5 * 1024 * 1024; // 5MB
+
 pub(crate) struct File(pub(crate) db::file::Blob);
 
 impl File {
     /// Gets file from database
     pub fn from_path(conn: &Connection, path: &str) -> Result<File> {
-        Ok(File(db::file::get_path(conn, path)?))
+        Ok(File(db::file::get_path(conn, path, MAX_FILE_SIZE)?))
```
Maybe we should use an environmental variable so we can adjust the limit on the fly instead of having to redeploy

pietroalbini replied:
Hmm, I don't think an environment variable is strictly needed, and implementing it correctly might be complex (I'd prefer not to read and parse it every single request).

Kixiron commented on src/storage/mod.rs:
```diff
@@ -163,7 +169,8 @@ impl<'a> Storage<'a> {
             })
             .map(|(file_path, file)| -> Result<_, Error> {
-                let (content, alg) = compress(file)?;
+                let alg = DEFAULT_COMPRESSION;
+                let content = compress(file, alg)?;
```
What's the purpose of setting a variable to `DEFAULT_COMPRESSION` instead of using it directly?

pietroalbini replied:
I put `DEFAULT_COMPRESSION` in a variable as that makes changing the function to accept the algorithm as a parameter easier in the future. I'm not too attached to that though, I can surely change this.

Kixiron commented on src/storage/mod.rs:
```diff
@@ -14,6 +14,7 @@ use std::io::Read;
 use std::path::{Path, PathBuf};
 
 const MAX_CONCURRENT_UPLOADS: usize = 1000;
+const DEFAULT_COMPRESSION: CompressionAlgorithm = CompressionAlgorithm::Zstd;
```
These could also be good targets for an environmental variable of some sort, we don't have multiple algos atm but in the future we might, and adjusting concurrent uploads would also be nice

pietroalbini replied:
I don't think that's required for the current PR. Feel free to send a future PR though.

Nemo157 replied:
With custom dictionaries the compression will probably be chosen on a per-file basis, so not easily configurable other than by changing code.

Kixiron commented on src/storage/mod.rs:
```diff
@@ -14,6 +14,7 @@ use std::io::Read;
 use std::path::{Path, PathBuf};
 
 const MAX_CONCURRENT_UPLOADS: usize = 1000;
```
Unrelated, but it occurred to me that we could potentially benefit from dynamically selecting the number of concurrent uploads based on the current conditions, e.g. decrease the number of uploads if we're under heavier load

Kixiron commented on src/storage/mod.rs:
```diff
@@ -341,7 +349,8 @@ mod test {
         let uploads: Vec<_> = (0..=MAX_CONCURRENT_UPLOADS + 1)
             .map(|i| {
-                let (content, alg) = compress("fn main() {}".as_bytes()).unwrap();
+                let alg = DEFAULT_COMPRESSION;
+                let content = compress("fn main() {}".as_bytes(), alg).unwrap();
```
Again it seems strange to set a variable to `DEFAULT_COMPRESSION`

pietroalbini replied:
See the other comment.

Kixiron commented on src/storage/mod.rs:
```diff
@@ -369,16 +378,20 @@ mod test {
+            };
+            test_roundtrip(std::slice::from_ref(&blob));
+            assert_eq!(decompress(data.as_slice(), *alg).unwrap(), orig.as_bytes());
+        }
```
This could also be a good place to make sure the buffering works (ensuring it works for compressed files as well as db ones)

pietroalbini replied:
Hmm? This is not testing buffering.

Kixiron replied:
Buffering is still involved within the pipeline though, isn't it?

pietroalbini replied:
Yeah, but the fact buffering works is already tested elsewhere. We don't need to add such tests every time we access the database.

Kixiron commented on src/storage/mod.rs:
```diff
@@ -390,7 +399,27 @@ mod test {
+                .and_then(|io| io.get_ref())
+                .and_then(|err| err.downcast_ref::<crate::error::SizeLimitReached>())
+                .is_some());
         }
```
Testing edge cases here as well would be good, like undersized and files that are on the buffer limit

pietroalbini replied:
Good point.

Kixiron commented on src/storage/s3.rs:
```diff
@@ -190,6 +191,47 @@ pub(crate) mod tests {
     }
 
+    #[test]
+    fn test_get_too_big() {
```
Buffer testing here as well?

pietroalbini replied:
What do you mean?

Kixiron replied:
Well, the sized buffer is used indirectly by the things being tested, and so making sure that the sized buffer preforms as expected in a real scenario seems logical to me

Kixiron commented on src/storage/database.rs:
```diff
@@ -16,18 +16,37 @@ impl<'a> DatabaseBackend<'a> {
+    pub(super) fn get(&self, path: &str, max_size: usize) -> Result<Blob, Error> {
         use std::convert::TryInto;
 
+        // The maximum size for a BYTEA (the type used for `content`) is 1GB, so this cast is safe:
+        // https://www.postgresql.org/message-id/162867790712200946i7ba8eb92v908ac595c0c35aee%40mail.gmail.com
+        let max_size = max_size.min(std::i32::MAX as usize) as i32;
```
Maybe add a debug assertion here to make sure that postgres behaves as expected

pietroalbini replied:
Hmm, I don't see a way of asserting what the postgres storage limit is. Do you have any idea that doesn't involve writing 1GB of data in the database?

Kixiron replied:
I meant asserting that the size we're given matches what we believe will happen, e.g `assert!(max_size <= i32::max_value() as usize);`

pietroalbini replied:
Well, the line we're commenting on already ensures that... Also, the tests will catch it if for whatever reason it stops working (I needed to add this line *because* tests were failing).

Kixiron commented on src/storage/database.rs:
```diff
@@ -91,22 +110,69 @@ mod tests {
-                backend.get("dir/foo.txt")?
+                backend.get("dir/foo.txt", std::usize::MAX)?
             );
 
             // Test that other files are not returned
             assert!(backend
-                .get("dir/bar.txt")
+                .get("dir/bar.txt", std::usize::MAX)
                 .unwrap_err()
                 .downcast_ref::<PathNotFoundError>()
                 .is_some());
             assert!(backend
-                .get("foo.txt")
+                .get("foo.txt", std::usize::MAX)
```
The usages of `usize::MAX` seem strange to me, it seems like testing with a sane-ish number is better and more realistic than this, which is essentially not having a limit at all

pietroalbini replied:
These tests are not testing the size limit, so there is no point in setting a lower limit.

Kixiron commented on src/storage/database.rs:
```diff
@@ -91,22 +110,69 @@ mod tests {
     }
+
+    #[test]
+    fn test_get_too_big() {
```
I know I've said it a lot, but maybe more buffering tests here?

pietroalbini replied:
What do you mean?

Nemo157 commented:
LGTM

pietroalbini commented:
Addressed all the review comments, and changed the maximum size to 50MB for non-HTML files.

I think this is ready to go, I'll follow up with a PR that does a couple refactorings and then adds a test for `File::from_path`. One last check @Kixiron @jyn514?

pietroalbini commented:
Nevermind, also added the final test.

jyn514 commented on src/storage/database.rs:
```diff
@@ -26,29 +26,28 @@ impl<'a> DatabaseBackend<'a> {
         } else {
             let row = rows.get(0);
+
+            if row.get::<_, bool>("is_too_big") {
```
nit: I think rust can infer the type here.
```suggestion
            if row.get("is_too_big") {
```

jyn514 commented on src/storage/mod.rs:
```diff
@@ -369,16 +387,60 @@ mod test {
+        let exact = &[b'A'; MAX_SIZE] as &[u8];
+        let big = &[b'A'; MAX_SIZE * 2] as &[u8];
+
+        for alg in CompressionAlgorithm::AVAILABLE {
+            let compressed_small = compress(small, *alg).unwrap();
+            let compressed_exact = compress(exact, *alg).unwrap();
+            let compressed_big = compress(big, *alg).unwrap();
```
nit: you don't need all this `*alg`:
```suggestion
        for &alg in CompressionAlgorithm::AVAILABLE {
            let compressed_small = compress(small, alg).unwrap();
            let compressed_exact = compress(exact, alg).unwrap();
            let compressed_big = compress(big, alg).unwrap();
```

jyn514 commented on src/bin/cratesfyi.rs:
```diff
@@ -1,17 +1,20 @@
     logger_init();
 
-    CommandLine::from_args().handle_args();
+    CommandLine::from_args().handle_args()?;
+    Ok(())
```
nit: I think this looks nicer.
```suggestion
    CommandLine::from_args().handle_args()
```

jyn514 commented on src/config.rs:
```diff
@@ -0,0 +1,37 @@
+}
+
+impl Config {
+    pub fn from_env() -> Result<Self, Error> {
```
The idea is that over time we'll move more of the environment variables here, instead of being spread out over the codebase? I like it, it makes things much easier to test as well.

pietroalbini replied:
Yep, that's the goal!

Kixiron replied:
I completely support that, having a concrete storage of env vars in one place would be fantastic

jyn514 commented:
Looks good to me modulo nits.

pietroalbini commented:
@jyn514 addressed the nits.

I also updated how tests can override configuration values to this:

```rust
env.override_config(|config| {
   config.key = value;
});
```

Using the closure allows to change the default configuration for tests in the future without having to update the base config in each test that overrides it.

jyn514 commented:
Looks great, thanks!
